src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentTypeParamDirective.cs

```diff
@@ -7,23 +7,45 @@ namespace Microsoft.AspNetCore.Razor.Language.Components
 {
     internal class ComponentTypeParamDirective
     {
-        public static readonly DirectiveDescriptor Directive = DirectiveDescriptor.CreateDirective(
-            "typeparam",
-            DirectiveKind.SingleLine,
-            builder =>
-            {
-                builder.AddMemberToken(ComponentResources.TypeParamDirective_Token_Name, ComponentResources.TypeParamDirective_Token_Description);
-                builder.Usage = DirectiveUsage.FileScopedMultipleOccurring;
-                builder.Description = ComponentResources.TypeParamDirective_Description;
-            });
+        public static DirectiveDescriptor Directive = null;
 
-        public static RazorProjectEngineBuilder Register(RazorProjectEngineBuilder builder)
+        public static RazorProjectEngineBuilder Register(RazorProjectEngineBuilder builder, bool supportConstraints)
         {
             if (builder == null)
             {
                 throw new ArgumentNullException(nameof(builder));
             }
 
+            if (Directive == null)
+            {
+                // Do nothing and assume the first registration wins. In real life this directive is only ever registered once.
+                if (supportConstraints)
+                {
+                    Directive = DirectiveDescriptor.CreateDirective(
+                        "typeparam",
```

Comment:
We normally don't do const for these things because they add an extra level of indirection and have very limited value since it's not used or referred to outside of the definition of this method. If you check all other directives they follow the same pattern as this one. I'm going to leave it as is to keep it consistent with the other directive definitions.